src/Engine/Contest/Solvers/ContestDijkstra.cpp

```diff
@@ -237,26 +240,34 @@ ContestDijkstra::AddEdges(const ScanTaskPoint origin,
        search */
     destination.SetPointIndex(first_finish_candidate);
 
+  const GeoPoint originGeo = GetPoint(origin).GetLocation();
   const unsigned weight = GetStageWeight(origin.GetStageNumber());
 
   bool previous_above = false;
   for (const ScanTaskPoint end(destination.GetStageNumber(), n_points);
        destination != end; destination.IncrementPointIndex()) {
     bool above = GetPoint(destination).GetIntegerAltitude() >= min_altitude;
 
-    if (above) {
-      const unsigned d = weight * CalcEdgeDistance(origin, destination);
-      Link(destination, origin, d);
-    } else if (previous_above) {
-      /* After excessive thinning, the exact TracePoint that matches
-         the required altitude difference may be gone, and the
-         calculated result becomes overly pessimistic.  This code path
-         makes it optimistic, by checking if the previous point
-         matches. */
-
-      /* TODO: interpolate the distance */
-      const unsigned d = weight * CalcEdgeDistance(origin, destination);
-      Link(destination, origin, d);
+    const unsigned distance = GetPoint(destination).DistanceTo(originGeo);
```

Comment:
This integer cast loses precision for no reason. And it is calculated even if no `min_distance` is configured, adding overhead to all implementations.
Do we really need this field in the base class?

Reply:
I moved this to the base class, but has to make `AddEdges` virtual as a result.
Is that acceptable to you?

Reply:
Your commit which makes it virtual adds a lot of duplicate code. That is all but acceptable.
I'm always cautious to make methods virtual, because the overhead this involves can sometimes be prohibitive (and sometimes doesn't matter), but in this case, the "virtual" is not the problem.
I suggest moving the "min_distance" feature to a separate commit (but leaving it on `class ContestDijkstra`), convert it to `double` to avoid implicit casts, and do the `DistanceTo()` call only if `min_distance>0`.

Reply:
Ok, will do. By separate commit you mean separate pull request?

Reply:
No, separate commit, same PR.
Check out [stgit](https://stacked-git.github.io/guides/tutorial/), it's a terrific tool which helps you manage PR branches in a proper and clean way, lets you move patches around, and helps avoid unclean fixup commits like the one you added to this PR.